src/openms/source/FORMAT/FASTAFile.cpp

```diff
@@ -39,179 +39,240 @@
 
-    infile_.open(filename.c_str(), std::ios::binary | std::ios::in);
+    bool FASTAFile::getLine_(std::istream& is, std::string& t){
+        t.clear();
+        std::istream::sentry se(is, true);
+        if (!se)
+        { // the stream has an error
+            return false;
+        }
+        std::streambuf* sb = is.rdbuf();
+        if(entries_read_== 0)//
+        {
+            if(sb->sgetc() == '>') return false;
+            if(sb->sgetc() == '#')
+            {
+                is.ignore(numeric_limits<streamsize>::max(),'\n');//skipping the following line without reading
+                return true;
+            }
+        }
+        for (;;)
+        {
+            int c = sb->sbumpc();// get and advance to next char
+            switch (c)
+            {
+                case '\n':
+                    if (sb->sgetc() == '>') // peek current char
+                    {
+                        return false;//reaching the beginning of the next protein-entry
+                    }
+                    return true;
+                case '\r':
+                    if (sb->sgetc() == '\n') // peek current char
+                    {
+                        sb->sbumpc(); // consume it
+                    }
+                    return true;
+                case ' ':
+                    break;
+                case '\t':
+                    break;
+                case std::streambuf::traits_type::eof():
+                    is.setstate(std::ios::eofbit); // still allows: while(is == true)
+                    if (t.empty())
+                    { // only if we just started a new line, we set the is.fail() == true, ie. is == false
+                        is.setstate(std::ios::badbit);
+                    }
+                    return false;
+                default:
+                    t += (char)c;
+            }
+        }
+    }
 
-    // Skip the header of PEFF files (http://www.psidev.info/peff)
-    std::string line;
-    std::streampos firstline = 0;
-    while (TextFile::getLine(infile_, line))
+    bool FASTAFile::readEntry_(std::string & id, std::string & seq)
     {
-      if (!line.empty() && line[0] != '#')
-      {
-        break;
-      }
-      firstline = infile_.tellg();
-    }
-    infile_.seekg(firstline);
-  
-    // automatically deletes old handles
-    reader_ = std::unique_ptr<void, std::function<void(void*) > >(new FASTARecordReader(infile_),
-      [](void* ptr)
-      { // lambda with custom cast
-      delete static_cast<FASTARecordReader*>(ptr);
-      });
-
-    entries_read_ = 0;
-  }
-
-  bool FASTAFile::readNext(FASTAEntry& protein)
-  {
-    if (seqan::atEnd(*static_cast<FASTARecordReader*>(reader_.get())))
-    { 
-      // do NOT close(), since we still might want to seek to certain positions
-      return false;
+       std::string line;
+       if(TextFile::getLine(infile_,line))//using Textfile::getLine to be able to read '>'
+       {
+           line.erase(0,1);
```

Comment:
expensive operation? Maybe its not a problem tough... Benchmark first. If the performance is bad, this may need some attention.
Maybe also check if '>' is the first character?